File: problems/problem 4/test_cases.py
import unittest, subprocess, os, sys

SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
SOLUTION_PATH = os.path.join(SCRIPT_DIR, "solution.py")

class TestQ4(unittest.TestCase):
    def run_solution(self, inp: str) -> str:
        result = subprocess.run(
            [sys.executable, SOLUTION_PATH],
            input=inp,
            capture_output=True,
            text=True,
            timeout=5
        )
        return result.stdout.strip()

    def test_samples(self):
        # Sample from PDF
        self.assertEqual(
            self.run_solution("3 3\n1 2 3\n4 5 6\n7 8 9\n"), "29"
        )

    def test_hidden(self):
        # Hidden test 1
        self.assertEqual(
            self.run_solution("2 2\n-1 2\n1 3\n"), "4"
        )
        # Hidden test 2 (PDF said 11, but true max is 12 ✅)
        self.assertEqual(
            self.run_solution("3 4\n1 2 3 4\n2 2 1 1\n5 1 1 1\n"), "12"
        )
        # Hidden test 3
        self.assertEqual(
            self.run_solution("4 4\n1 1 1 1\n1 1 1 1\n1 1 1 1\n1 1 1 9\n"), "15"
        )

if __name__ == "__main__":
    unittest.main()

File: problems/problem 4/test_cases.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int maxPathSum(int** grid, int m, int n);

int failures = 0;

void check(int **grid, int m, int n, int expected) {
    int result = maxPathSum(grid, m, n);
    if (result == expected) {
        printf("PASS\n");
    } else {
        printf("FAIL Expected=%d Got=%d\n", expected, result);
        failures++;
    }
}

int main() {
    // Sample test
    int g1_data[3][3] = {{1,2,3},{4,5,6},{7,8,9}};
    int* g1[3] = {g1_data[0], g1_data[1], g1_data[2]};
    check(g1, 3, 3, 29);

    // Hidden test 1
    int g2_data[2][2] = {{-1,2},{1,3}};
    int* g2[2] = {g2_data[0], g2_data[1]};
    check(g2, 2, 2, 4);

    // Hidden test 2 (PDF says 11, but real max path is 12 ✅)
    int g3_data[3][4] = {{1,2,3,4},{2,2,1,1},{5,1,1,1}};
    int* g3[3] = {g3_data[0], g3_data[1], g3_data[2]};
    check(g3, 3, 4, 12);

    // Hidden test 3
    int g4_data[4][4] = {
        {1,1,1,1},
        {1,1,1,1},
        {1,1,1,1},
        {1,1,1,9}
    };
    int* g4[4] = {g4_data[0], g4_data[1], g4_data[2], g4_data[3]};
    check(g4, 4, 4, 15);

    if (failures > 0) return 1;
    return 0;
}
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int maxPathSum(int** grid, int m, int n);

int failures = 0;

void check(int **grid, int m, int n, int expected) {
    int result = maxPathSum(grid, m, n);
    if (result == expected) {
        printf("PASS\n");
    } else {
        printf("FAIL Expected=%d Got=%d\n", expected, result);
        failures++;
    }
}

int main() {
    // Sample test
    int g1_data[3][3] = {{1,2,3},{4,5,6},{7,8,9}};
    int* g1[3] = {g1_data[0], g1_data[1], g1_data[2]};
    check(g1, 3, 3, 29);

    // Hidden test 1
    int g2_data[2][2] = {{-1,2},{1,3}};
    int* g2[2] = {g2_data[0], g2_data[1]};
    check(g2, 2, 2, 4);

    // Hidden test 2 (PDF says 11, but real max path is 12 ✅)
    int g3_data[3][4] = {{1,2,3,4},{2,2,1,1},{5,1,1,1}};
    int* g3[3] = {g3_data[0], g3_data[1], g3_data[2]};
    check(g3, 3, 4, 12);

    // Hidden test 3
    int g4_data[4][4] = {
        {1,1,1,1},
        {1,1,1,1},
        {1,1,1,1},
        {1,1,1,9}
    };
    int* g4[4] = {g4_data[0], g4_data[1], g4_data[2], g4_data[3]};
    check(g4, 4, 4, 15);

    if (failures > 0) return 1;
    return 0;
}
